### ISSUE TITLE
What does onBeforeShow get us that onRender and onAttach do not provide?

### ISSUE THREAD

**Issue by MeoMix** [marionettejs/backbone.marionette#2539]
I feel like onBeforeShow had a place in Marionette before the introduction of onAttach. Now, though, I'm not so sure.

Is anyone aware of a scenario that onBeforeShow would provide useful functionality?

**Comment by jasonLaster** [comment]
Just make sure I'm on the same page, onBeforeAttatch could be used?

**Comment by MeoMix** [comment]
Well, at least in all of my workflow, I've always used 'onRender' on the assumption that the view was then about to be shown, or already is shown, on the page.

The 'show' event handler was nice before 'attach' existed because it provided a mechanism for (albeit sometimes poorly..) detecting that an element was now apart of the DOM.

With the advent of 'attach' I find myself struggling for uses for 'show' except in scenarios where Marionette should trigger an attach event, but does not (i.e. in CollectionView re-rendering).

It would be really helpful to me if there were use cases outlined for each of the hooks. They could be really contrived, but I struggle to tell users on Gitter why they're experiencing problems following the docs suggestion of 'onBeforeShow'. They often then call 'render' and expect children views to be re-rendered, but 'onBeforeShow' doesn't trigger in that scenario.

**Comment by paulfalgout** [comment]
Personally, I would support essentially using `onBeforeAttach` and `onAttach` to replace previous uses for `onBeforeShow` and `onShow` where `before:attach` and `attach` would be expected to bubble down through all of the children also being attached.  And then show would only be triggered on the specific view being shown in the region, with possibly no reason for the event bubbling.  Along with this `before:show` would need to be the first thing to happen on a `show` where-as currently it is waiting for the layout to `render` for the supposed perf boosts.  `onBeforeAttach` should be used there.

*TL:DR* make show only related to the view being shown in a region.  Its realistic use-case is limited.  Use `attach` for everything else. 
Similarly, this is loosely related to https://github.com/marionettejs/backbone.marionette/issues/2299

**Comment by paulfalgout** [comment]
To answer your initial question..  `onBeforeShow` just like the newer `onBeforeAttach` currently allows you to render childViews into a layout during a `show` as the when you call show the view is rendered then `onBeforeShow` is called.. so essentially:
`show => onBeforeRender -> onRender -> onBeforeShow -> onBeforeAttach -> [view is added to DOM] onAttach -> onShow`
As mentioned previously I think it makes more sense to move `onBeforeShow` so that the event order is
`show => onBeforeShow -> onBeforeRender -> onRender -> onBeforeAttach -> [view is added to DOM] onAttach -> onShow`

**Comment by MeoMix** [comment]
I think I'm still struggling to understand the usefulness of the 'show' hooks.

To me, there are only a few phases in the initial lifecycle of showing a view:

* View is about to be rendered. Do any pre-processing of templateHelpers/etc. here.
* View is rendered. Template has been injected into a wrapping tag. Element has not been attached to the DOM. You are free to work with it as you please, but height/width/etc. are unknown.
* View is attached to the DOM. You are free to measure its dimensions and query it in all the ways you see fit.

I can also imagine that there are uses for onBeforeAttach and that I just have yet to encounter them. When I render a view I'm expecting it to be attached to the DOM a moment later, or already attached. In either case, I'm doing the work I need to decorate it in onRender or onAttach since the state of the view's element relative to the DOM doesn't change at all between onRender and onAttach.

The concept of a view being 'shown' in a region doesn't have any importance to me. Why would the child care that it's being rendered and attached to the DOM via a region versus manually? I think I'd back down from this very quickly if someone could show me a use case for a 'show' event which isn't able to be handled exactly the same with render/attach events.

**Comment by paulfalgout** [comment]
The usefulness of the show event in my mind is legacy.  Attach hasn't existed before, so show was used like attach.  I don't think show is all that important in light of attach, but there are current issues with attach #2546   You would use onBeforeAttach if you want to do something to the rendered view before it is attached to the dom..  
say you have a view with regions and you want to put something into those regions.. you have two choices.
you can render that view, place it in the dom, then once it is in the dom add the children to it.
Or you can render the view, add the children to it, and the place it in the dom..  (theoretically faster, but not very well backed up)  onBeforeAttach would solve this issue.  Currently onBeforeShow is recommended, however that is because, I believe it is in the wrong place https://github.com/marionettejs/backbone.marionette/pull/2552

**Comment by MeoMix** [comment]
Okay, we are in agreement regarding show. :)

The usefulness you describe is work I do in onRender. The view has already been rendered when onRender fires so it is safe to render child views at this point in time. Is there any reason why I should not be doing that work in onRender? It has the added benefit of re-rendering child views if the parent view calls render on itself.

**Comment by jiaweihli** [comment]
Is this part of the documentation still accurate?: http://marionettejs.com/docs/marionette.layoutview.html#efficient-nested-view-structures

Something I find very tedious re: `onBeforeShow` is (as @MeoMix mentioned) that it isn't called during a `render()` call.  This means I have to remember to manually call it.

**Comment by ianmstew** [comment]
Hey all, a new, better official recommendation for rendering child views will be in v3.

**Comment by MeoMix** [comment]
That's great to hear :D

**Comment by paulfalgout** [comment]
Closed by https://github.com/marionettejs/backbone.marionette/pull/2850
